component/kafka/simple/consumer.go

```diff
@@ -0,0 +1,300 @@
+	}
+
+	return c.consumePartitionUnit(ctx, ch, partition)
+}
+
+func (c *consumerHandler) startingOffsetAfterLatestOffset(latestOffset int64, partition int32) bool {
+	return c.component.startingOffsets[partition] >= latestOffset
+}
+
+func (c *consumerHandler) partitionHasReachedLatestOffset(partition int32) {
+	delete(c.partitionsWithLatestOffsetUnreached, partition)
+	if len(c.partitionsWithLatestOffsetUnreached) == 0 {
+		c.notificationAlreadySent = true
+		// As the consumer can be retried, we have to make sure the channel is closed only once.
+		c.component.once.Do(func() {
+			close(c.component.latestOffsetReachedChan)
+		})
+	}
+}
+
+func (c *consumerHandler) consumePartitionUnit(ctx context.Context, ch <-chan *sarama.ConsumerMessage, partition int32) error {
+	for {
+		select {
+		case msg, ok := <-ch:
+			if !ok {
+				log.Debug("messages channel closed")
+				return nil
+			}
+			log.Debugf("message claimed: value = %s, timestamp = %v, partition = %d topic = %s", string(msg.Value), msg.Timestamp, msg.Partition, msg.Topic)
+			messageStatusCountInc(messageReceived, partition, msg.Topic)
+			err := c.unit(ctx, msg)
+			if err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			if ctx.Err() != context.Canceled {
+				log.Infof("closing consumer: %v", ctx)
+			}
+			return nil
+		}
+	}
+}
+
+func (c *consumerHandler) consumePartitionBatch(ctx context.Context, ch <-chan *sarama.ConsumerMessage, partition int32) (err error) {
+	// We need to enforce flushing the buffer before to return (e.g., context cancelled but messages are still buffered)
+	defer func() {
+		errFlush := c.flush(ctx)
+		if errFlush != nil {
+			err = errFlush
+		}
+	}()
+
+	for {
+		select {
+		case msg, ok := <-ch:
+			if ok {
+				log.Debugf("message claimed: value = %s, timestamp = %v, partition = %d topic = %s", string(msg.Value), msg.Timestamp, msg.Partition, msg.Topic)
+				messageStatusCountInc(messageReceived, partition, msg.Topic)
+				err := c.insertMessage(ctx, msg)
+				if err != nil {
+					return err
+				}
+			} else {
+				log.Debug("messages channel closed")
+				return nil
+			}
+		case <-c.ticker.C:
+			err := c.flush(ctx)
+			if err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			if ctx.Err() != context.Canceled {
+				log.Infof("closing consumer: %v", ctx)
+			}
+			return nil
+		}
+	}
+}
+
+func (c *consumerHandler) unit(ctx context.Context, msg *sarama.ConsumerMessage) error {
+	messageStatusCountInc(messageProcessed, msg.Partition, msg.Topic)
+	ctx, sp := c.getContextWithCorrelation(ctx, msg)
+	messages := []kafka.Message{kafka.NewMessage(ctx, sp, msg)}
+
+	btc := kafka.NewBatch(messages)
+	err := c.component.proc(btc)
+	if err != nil {
+		if ctx.Err() == context.Canceled {
+			return fmt.Errorf("context was cancelled after processing error: %w", err)
+		}
+		err = c.executeFailureStrategy(messages, err)
+		if err != nil {
+			return err
+		}
+	} else {
+		c.mu.Lock()
+		c.hasProcessedMessages = true
```

Comment:
Mutex locks (`c.mu`) are held when executing the failure strategy or when message consumption was successful, but not when running the processor function nor interacting with the Kafka server. I wonder: why is locking necessary? e.g. what concurrent operations are happening that you would like to safeguard against overwriting the same fields/state?